inst/include/dust/r/random.hpp

```diff
@@ -40,6 +53,113 @@ std::vector<typename rng_state_type::int_type> as_rng_seed(cpp11::sexp r_seed) {
   return seed;
 }
 
+namespace {
+
+template<typename T>
+std::string algorithm_name() {
+  std::string ret;
+  if (std::is_same<T, xoshiro128plus_state>::value) {
+    ret = "xoshiro128plus";
+  } else if (std::is_same<T, xoshiro128plusplus_state>::value) {
+    ret = "xoshiro128plusplus";
+  } else if (std::is_same<T, xoshiro128starstar_state>::value) {
+    ret = "xoshiro128starstar";
+  } else if (std::is_same<T, xoroshiro128plus_state>::value) {
+    ret = "xoroshiro128plus";
+  } else if (std::is_same<T, xoroshiro128plusplus_state>::value) {
+    ret = "xoroshiro128plusplus";
+  } else if (std::is_same<T, xoroshiro128starstar_state>::value) {
+    ret = "xoroshiro128starstar";
+  } else if (std::is_same<T, xoshiro256plus_state>::value) {
+    ret = "xoshiro256plus";
+  } else if (std::is_same<T, xoshiro256plusplus_state>::value) {
+    ret = "xoshiro256plusplus";
+  } else if (std::is_same<T, xoshiro256starstar_state>::value) {
+    ret = "xoshiro256starstar";
+  } else if (std::is_same<T, xoshiro512plus_state>::value) {
+    ret = "xoshiro512plus";
+  } else if (std::is_same<T, xoshiro512plusplus_state>::value) {
+    ret = "xoshiro512plusplus";
+  } else if (std::is_same<T, xoshiro512starstar_state>::value) {
+    ret = "xoshiro512starstar";
+  }
+  return ret;
+}
+
+template <typename rng_state_type>
+cpp11::raws rng_state_vector(prng<rng_state_type>* rng) {
+  auto state = rng->export_state();
+  size_t len = sizeof(typename rng_state_type::int_type) * state.size();
+  cpp11::writable::raws r_state(len);
+  std::memcpy(RAW(r_state), state.data(), len);
+  return r_state;
+}
+
+}
+
+template <typename rng_state_type>
+SEXP rng_pointer_init(int n_streams, cpp11::sexp r_seed) {
+  auto seed = as_rng_seed<rng_state_type>(r_seed);
+  auto *rng = new prng<rng_state_type>(n_streams, seed);
+  auto r_ptr = cpp11::external_pointer<prng<rng_state_type>>(rng);
+  auto r_state = rng_state_vector(rng);
+  return cpp11::writable::list({r_ptr, r_state});
+}
+
+// Start with the assumption that we'll pass in the R6 object, might
+// write a simpler version later.
+template <typename rng_state_type>
+prng<rng_state_type>* rng_pointer_get(cpp11::environment obj,
```

Comment:
Possible useful reference for #271. If we wanted to add this and support saving and loading, see the serialisation examples:
https://pybind11.readthedocs.io/en/stable/advanced/classes.html#pickling-support
https://github.com/pybind/pybind11/blob/master/tests/test_pickling.cpp

Reply:
The R serialisation has no hooking functionality, it's super annoying